core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/AllocationDecider.java

```diff
@@ -98,4 +98,15 @@ public Decision canAllocate(RoutingNode node, RoutingAllocation allocation) {
     public Decision canRebalance(RoutingAllocation allocation) {
         return Decision.ALWAYS;
     }
+
+    /**
+     * Returns a {@link Decision} whether the given primary shard can be
+     * forcibly allocated on the given node. This method should only be called
+     * for unassigned primary shards where the node has a shard copy on disk.
+     */
+    public Decision canForceAllocatePrimary(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+        assert shardRouting.primary() : "must not call canForceAllocatePrimary on a non-primary shard " + shardRouting;
+        assert shardRouting.unassigned() : "must not call canForceAllocatePrimary on an assigned shard " + shardRouting;
+        return Decision.YES; // by default, a decider will allow force allocation of the primary
```

Comment:
My concern with your suggestion (which was the initial approach) is that for any decider that overrides `canForceAllocatePrimary`, it is easy for it to forget to take into account the decision of `canAllocate`.  That's why I figured it would be better to put the invocation of `canAllocate` in `AllocationDeciders`, then if the decision is `NO` (as opposed to throttle, for example), then call the deciders `canForceAllocatePrimary` implementation.  The only downside to this that I can think of is that its rigid in that no `canForceAllocatePrimary` method would be able to override a non-NO `canAllocate` decision - but I don't see any reason why we would want that?